core/parameters/src/config.rs

```diff
@@ -241,3 +252,30 @@ pub struct WitnessConfig {
     /// Soft size limit of storage proof used to validate new transactions in ChunkStateWitness.
     pub new_transactions_validation_state_size_soft_limit: usize,
 }
+
+/// Configures gas limit adjustments at runtime.
+///
+/// This functionality is not intended for production, but only for tests or benchmarks.
+///
+/// A possible use case is making a node that runs benchmarks automatically set the maximum gas
+/// limit that it can sustain.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct GasLimitAdjustmentConfig {
+    /// Override for `GAS_LIMIT_ADJUSTMENT_FACTOR`.
+    ///
+    /// The default adjustment factor might not be effective enough.
+    pub adjustment_factor_override: u64,
+    /// The maximum chunk apply time considered sustainable, in milliseconds.
+    pub max_chunk_apply_time: u64,
```

Comment:
This and 2 other time values below are unlikely to become the subject to consensus rules as they are non-deterministic and hard to enforce. So they could possibly be a part of the node config - e.g. it makes sense for them to be changed during operation without rebuilding the `neard`.
I do realize though that this could be tricky to wire these values from there to the place where they are used. So unless @Longarithm and @robin-near have any suggestions, we can leave them here and add a comment about this.

Reply:
My concern about moving some parameters into node config would be that IMO with configuration split across `RuntimeConfig` and `ClientConfig` it becomes significantly harder to reason about and maintain dynamic gas limit adjustments. Anyway, [this draft](https://github.com/near/nearcore/compare/master...mooori:nearcore:client-config-gas-limit?expand=1) gives an idea of the wiring needed to put it into `ClientConfig`.

For now this feature is supposed to be used only in tests and benchmarks:

- Testing usually involves compilation, so having everything in `RuntimeConfig` might be fine for this use case.
- I agree that moving some of the parameters into node config could facilitate running benchmarks. Though for a given setup I would expect the time parameters to be relatively constant. Then we might be fine with having them in `RuntimeConfig` and ocassionally adjust them in the benchmark setup, e.g. [here](https://github.com/near/nearcore/blob/3bdc50d2a1c537bd300ce0c0df2813cd574c3ee9/core/parameters/src/config_store.rs#L144).

Reply:
Runtime config is a great place for `adjustment_factor_override`; however, `_time` parameters indeed doesn't define consensus rules. I think there is a compromise. `NightshadeRuntime` should be stateless, so introducing set of new parameters is messy. But we can move these parameters to `Chain`. The sequence of passing parameters could look as follows:

RuntimeConfig will have just a single `adjustment_factor_override: u64`
GasLimitAdjustmentConfig will have time-based parameters

* Client will take GasLimitAdjustmentConfig from ClientConfig
* -> ChainConfig in Chain::new
* -> Chain::gas_limit_adjustment_config field
* -> process_shard_update with new argument gas_limit_adjustment_config. That's probably fine because only the job created there will know the apply chunk time, based on which adjustment will happen
* -> apply_new_chunk with new argument
* -> NewChunkResult

I don't see it as completely necessary, because this is benchmark-only. However, doing it in client config may allow to use it on real situation if it ever comes to that; every node could configure its own adjustment parameters.